lib/duo.go

```diff
@@ -79,25 +127,112 @@ func (d *DuoClient) ChallengeU2f() (err error) {
 		return
 	}
 
-	_, err = d.DoStatus(txid, sid)
+	auth, status, err = d.DoStatus(txid, sid)
 	if err != nil {
 		return
 	}
 
-	// This one should block untile 2fa completed
-	auth, err = d.DoStatus(txid, sid)
-	if err != nil {
-		return
+	if status.Response.StatusCode == "u2f_sent" {
+		var response *u2fhost.AuthenticateResponse
+		allDevices := u2fhost.Devices()
+		// Filter only the devices that can be opened.
+		openDevices := []u2fhost.Device{}
+		for i, device := range allDevices {
+			err := device.Open()
+			if err == nil {
+				openDevices = append(openDevices, allDevices[i])
+				defer func(i int) {
+					allDevices[i].Close()
+				}(i)
+			}
+		}
+		if len(openDevices) == 0 {
+			return errors.New("no open u2f devices")
+		}
+
+		var (
+			err error
+		)
+		prompted := false
+		timeout := time.After(time.Second * 25)
+		interval := time.NewTicker(time.Millisecond * 250)
+		facet := ""
+		facetResponse, err := d.getTrustedFacet(status.Response.U2FSignRequest[0].AppID)
+		if err != nil {
+			return fmt.Errorf("failed to get trusted facets for u2f device. Err: %s", err)
+		}
+		// TODO: do more error checks
+		facet = facetResponse.TrustedFacets[0].Ids[0]
```

Comment:
Maybe this is what you meant by the TODO, but shouldn't we iterate through `TrustedFacets` to check that `appId` is in there? Dunno if it'll always be 0